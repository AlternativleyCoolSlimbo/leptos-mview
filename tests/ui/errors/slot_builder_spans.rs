//! Testing that there are no errors that cause the entire macro to error (i.e.
//! call-site error).
//! 
//! This file is for testing on the slot itself, see `com_builder_spans` for testing on components.

use leptos::*;
use leptos_mview::mview;

#[slot]
struct SChildren {
    an_attr: i32,
    children: ChildrenFn,
}

#[component]
fn TakesSChildren(s_children: SChildren) -> impl IntoView { let _ = s_children; }

fn missing_args() {
    _ = mview! {
        TakesSChildren {
            slot:SChildren { "hi" }
        }
    };
}

fn incorrect_arg_value() {
    _ = mview! {
        TakesSChildren { slot:SChildren an_attr="no" { "what" } }
    };
}

fn incorrect_closure_to_children() {
    let s = String::new();
    _ = mview! {
        TakesSChildren {
            slot:SChildren an_attr=1 |s| { "this is " {s} }
        }
    };
}

#[slot]
struct SNoChildren {
    an_attr: i32,
}

#[component]
fn TakesSNoChildren(s_no_children: SNoChildren) -> impl IntoView { let _ = s_no_children; }

fn incorrect_children() {
    _ = mview! {
        TakesSNoChildren {
            slot:SNoChildren an_attr=5 { "hey!" }
        }
    };
}

#[slot]
struct SClosureChildren {
    children: Callback<i32, View>,
}

#[component]
fn TakesSClosureChildren(s_closure_children: SClosureChildren) -> impl IntoView { let _ = s_closure_children; }

fn missing_closure_to_children() {
    _ = mview! {
        TakesSClosureChildren {
            slot:SClosureChildren { "hey!" }
        }
    };
}


fn main() {}
