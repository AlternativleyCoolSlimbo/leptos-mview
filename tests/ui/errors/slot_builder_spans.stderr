warning: use of deprecated method `SChildrenBuilder::<((), __children)>::build`: Missing required field an_attr
  --> tests/ui/errors/slot_builder_spans.rs:22:18
   |
22 |             slot:SChildren { "hi" }
   |                  ^^^^^^^^^
   |
   = note: `#[warn(deprecated)]` on by default

error[E0061]: this method takes 1 argument but 0 arguments were supplied
  --> tests/ui/errors/slot_builder_spans.rs:22:18
   |
22 |             slot:SChildren { "hi" }
   |                  ^^^^^^^^^ an argument of type `SChildrenBuilder_Error_Missing_required_field_an_attr` is missing
   |
note: method defined here
  --> tests/ui/errors/slot_builder_spans.rs:10:1
   |
10 | #[slot]
   | ^^^^^^^
   = note: this error originates in the derive macro `::leptos::typed_builder_macro::TypedBuilder` (in Nightly builds, run with -Z macro-backtrace for more info)
help: provide the argument
   |
22 |             slot:SChildren(/* SChildrenBuilder_Error_Missing_required_field_an_attr */) { "hi" }
   |                           +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

error[E0308]: mismatched types
  --> tests/ui/errors/slot_builder_spans.rs:29:49
   |
29 |         TakesSChildren { slot:SChildren an_attr="no" { "what" } }
   |                                         ------- ^^^^ expected `i32`, found `&str`
   |                                         |
   |                                         arguments to this method are incorrect
   |
note: method defined here
  --> tests/ui/errors/slot_builder_spans.rs:12:5
   |
12 |     an_attr: i32,
   |     ^^^^^^^-----

error[E0308]: mismatched types
  --> tests/ui/errors/slot_builder_spans.rs:37:38
   |
37 |             slot:SChildren an_attr=1 |s| { "this is " {s} }
   |                                      ^^^ expected `Rc<dyn Fn() -> Fragment>`, found closure
   |
   = note: expected struct `Rc<(dyn std::ops::Fn() -> Fragment + 'static)>`
             found closure `{closure@$DIR/tests/ui/errors/slot_builder_spans.rs:37:38: 37:41}`

error[E0599]: no method named `children` found for struct `SNoChildrenBuilder` in the current scope
  --> tests/ui/errors/slot_builder_spans.rs:53:42
   |
42 |   #[slot]
   |   ------- method `children` not found for this struct
...
51 |       _ = mview! {
   |  _________-
52 | |         TakesSNoChildren {
53 | |             slot:SNoChildren an_attr=5 { "hey!" }
   | |                                         -^^^^^^ method not found in `SNoChildrenBuilder<((i32,),)>`
   | |_________________________________________|
   |

error[E0277]: the trait bound `leptos::Callback<i32, leptos::View>: ToChildren<{closure@$DIR/tests/ui/errors/slot_builder_spans.rs:73:37: 73:43}>` is not satisfied
  --> tests/ui/errors/slot_builder_spans.rs:71:9
   |
71 |       _ = mview! {
   |  _________^
72 | |         TakesSClosureChildren {
73 | |             slot:SClosureChildren { "hey!" }
74 | |         }
75 | |     };
   | |_____^ the trait `ToChildren<{closure@$DIR/tests/ui/errors/slot_builder_spans.rs:73:37: 73:43}>` is not implemented for `leptos::Callback<i32, leptos::View>`
   |
   = help: the following other types implement trait `ToChildren<F>`:
             Box<(dyn FnMut() -> Fragment + 'static)>
             Box<(dyn FnOnce() -> Fragment + 'static)>
             Box<(dyn std::ops::Fn() -> Fragment + 'static)>
             Rc<(dyn std::ops::Fn() -> Fragment + 'static)>
   = note: this error originates in the macro `mview` (in Nightly builds, run with -Z macro-backtrace for more info)
